### DISCUSSION TITLE
System for testing migrations and seeds with Kysely on Vercel with PostgreSQL?

### DISCUSSION THREAD

**Discussion by lancejpollard** [vercel/community#2441] [Help]
### Summary

How do you recommend developing/testing that migrations will work and that seeds will not error out halfway through running them (because of some logic error let's say)? With Kysely and PostgreSQL.

I have a `./migrations` folder with a `./scripts/migrate.ts` for running migrations:

```ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import * as path from 'path'
import { Pool } from 'pg'
import { promises as fs } from 'fs'
import {
  Kysely,
  Migrator,
  PostgresDialect,
  FileMigrationProvider,
} from 'kysely'
import 'dotenv/config'

async function migrateToLatest() {
  const db = new Kysely<any>({
    dialect: new PostgresDialect({
      pool: new Pool({
        connectionString: process.env.DATABASE_URL,
      }),
    }),
  })

  const migrator = new Migrator({
    db,
    provider: new FileMigrationProvider({
      fs,
      migrationFolder: path.resolve('./migrations'),
      path,
    }),
  })

  const { error, results } = await migrator.migrateToLatest()

  results?.forEach(it => {
    if (it.status === 'Success') {
      console.log(
        `migration "${it.migrationName}" was executed successfully`,
      )
    } else if (it.status === 'Error') {
      console.error(`failed to execute migration "${it.migrationName}"`)
    }
  })

  if (error) {
    console.error('failed to migrate')
    console.error(error)
    process.exit(1)
  }

  await db.destroy()
}

migrateToLatest()
```

I then have my `@vercel/postgres-kysely` configuration:

```
import 'dotenv/config'
import { createKysely } from '@vercel/postgres-kysely'
import { DB } from 'kysely-codegen'

export const db = createKysely<DB>()
export { sql } from 'kysely'
```

As you can see, currently my migrations aren't going through the vercel configuration, should I change that?

Also, I am running:

```
yarn kysely-codegen --dialect postgres
```

which requires the `DATABASE_URL` environment variable, so I copied my _production_ `POSTGRES_URL` there, and it worked (running the migrations on Vercel).

Is there a better way to be configured?

Then my main concern: how should one be testing / developing the PostgreSQL schema and seeds? Should I have 2 production databases in my app, and one is for testing? If so, is that possible (not sure how I would do that, given all the PostgreSQL env vars were created for me)?

So say I get that working and have 2 dbs, run the migration on the test one, and some seeds, and it passes. Then I just run it on the production one, no problem. Okay I think that makes sense. Is there anything else that should be taken into consideration?

Is there a way to develop against a local PostgreSQL, in my Next.js app? And then when deployed it uses the production one? I wasn't able to use the `@vercel/postgres-kysely` code against my local PostgreSQL :/

Thanks for your help!

**Comment by sym233** [comment]
`@vercel/postgres-kysely` cannot connect local db, so I use `kysely` alone for development alone.
